File: DingRTC/Server/go/utils/signutils.go
package utils

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
)

const (
	AlgorithmHmacSHA256 = "sha256"
	IntLength           = 4
)

func GenerateSign(secretKey string, timestamp int32, salt int32) ([]byte, error) {
	timestampBytes := make([]byte, IntLength)
	binary.BigEndian.PutUint32(timestampBytes, uint32(timestamp))

	hash1 := hmac.New(sha256.New, timestampBytes)
	if _, err := hash1.Write([]byte(secretKey)); err != nil {
		return nil, err
	}

	signingT := hash1.Sum(nil)

	saltBytes := make([]byte, IntLength)
	binary.BigEndian.PutUint32(saltBytes, uint32(salt))

	hash2 := hmac.New(sha256.New, saltBytes)
	if _, err := hash2.Write(signingT); err != nil {
		return nil, err
	}

	signingS := hash2.Sum(nil)
	return signingS, nil

}

func Sign(key []byte, data []byte) ([]byte, error) {
	hash := hmac.New(sha256.New, key)
	if _, err := hash.Write(data); err != nil {
		return nil, err
	}

	return hash.Sum(nil), nil
}

File: DingRTC/Server/php/sign_utils.php
<?php
function generate_sign(string $key, int $timestamp, int $salt): string
{
    $buf_ts = pack('N', $timestamp);
    $sign_key = hash_hmac('sha256', $key, $buf_ts, true);

    $buf_salt = pack('N', $salt);
    $tmp_sign =  hash_hmac('sha256', $sign_key, $buf_salt, true);
    return $tmp_sign;
}

function sign(string $key, string $data): string
{
    return hash_hmac('sha256', $data, $key, true);
}
?>